src/mo_frontend/effect.ml

```diff
@@ -25,13 +25,32 @@ let typ phrase = phrase.note.note_typ
 let eff phrase = phrase.note.note_eff
 
 let is_triv phrase  =
-    eff phrase = T.Triv
+  eff phrase = T.Triv
+
+let is_shared_func exp =
+  T.(match typ exp with
+    | Func (Shared _, _, _, _, _) -> true
+    | _ -> false)
+
+let is_local_async_func exp =
+ T.(match typ exp with
+   | Func (Local, Returns,
+      { sort = Scope; _ }::_,
+       _,
+      [Async (Fut, Var (_ ,0), _)]) -> true
+   | _ -> false)
+
+let is_async_call exp1 inst exp2 =
+   is_shared_func exp1 ||
+   is_local_async_func exp1
 
 let effect_exp (exp:Syntax.exp) : T.eff = eff exp
 
 (* infer the effect of an expression, assuming all sub-expressions are correctly effect-annotated es*)
 let rec infer_effect_exp (exp:Syntax.exp) : T.eff =
   match exp.it with
+  | CallE (exp1, inst, exp2) when is_async_call exp1 inst exp2 ->
+    T.Await
```

Comment:
So calls to private `async` functions can throw now even if both `exp*` are `Triv`.

Reply:
Yes. As desired, because the application itself can throw, even if the function and arg are trivial. This is the usual unleashing of effects at application.